rbc/structure_type.py

```diff
@@ -0,0 +1,205 @@
+import operator
+from rbc.utils import get_version
+from llvmlite import ir
+from rbc.typesystem import Type
+
+if get_version('numba') >= (0, 49):
+    from numba.core import datamodel, extending, types, imputils, typing, cgutils, typeconv
+else:
+    from numba import datamodel, extending, types, typing, typeconv
+
+from numba.core.typing.templates import Registry
+
+typing_registry = Registry()
+lowering_registry = imputils.Registry()
+
+int8_t = ir.IntType(8)
+int32_t = ir.IntType(32)
+int64_t = ir.IntType(64)
+void_t = ir.VoidType()
+fp32 = ir.FloatType()
+fp64 = ir.DoubleType()
+
+
+def make_numba_struct(name, members, base=None, origin=None, _cache={}):
+    """Create numba struct type instance.
+    """
+    t = _cache.get(name)
+    if t is None:
+        if base is None:
+            base = Type.aliases.get('StructureNumbaType', StructureNumbaType)
+        assert issubclass(base, types.Type)  # base must be numba.types.Type
+
+        def model__init__(self, dmm, fe_type):
+            datamodel.StructModel.__init__(self, dmm, fe_type, members)
+
+        struct_model = type(name+'Model',
+                            (datamodel.StructModel,),
+                            dict(__init__=model__init__))
+        struct_type = type(name+'Type', (base,),
+                           dict(members=[t for n, t in members], origin=origin,
+                                __typesystem_type__=origin))
+        datamodel.registry.register_default(struct_type)(struct_model)
+        _cache[name] = t = struct_type(name)
+
+        tptr = StructureNumbaPointerType(t)
+        typeconv.rules.default_type_manager.set_compatible(
+            tptr, types.voidptr, typeconv.Conversion.safe)
+        typeconv.rules.default_type_manager.set_compatible(
+            types.voidptr, tptr, typeconv.Conversion.safe)
+
+    return t
+
+
+class StructureNumbaType(types.Type):
+    # pass_by_value = True
+    # return_as_first_argument = True  # use it only when deriving from types.PointerType
+
+    pass
+
+
+class StructureNumbaPointerType(types.Type):
+    """We are not deriving from CPointer because we may want to use
+    getitem for other purposes.
+    """
+    # pass_by_value = True
+    # return_as_first_argument = True  # use it only when deriving from types.PointerType
+    @property
+    def __typesystem_type__(self):
+        return self.dtype.origin.pointer()
+
+    def __init__(self, dtype):
+        self.dtype = dtype    # struct dtype
+        name = "(%s)*" % dtype
+        super().__init__(name)
+
+    @property
+    def key(self):
+        return self.dtype
+
+
+@datamodel.register_default(StructureNumbaPointerType)
+class StructureNumbaPointerTypeModel(datamodel.models.PointerModel):
+    pass
+
+
+@extending.infer_getattr
```

Comment:
Use `Registry.infer_getattr`

Reply:
It turns out that this is `typing_registry.register_attr`.
Fixed.